adam/learner/affordances.py

```diff
@@ -0,0 +1,195 @@
+    ObjectSemanticNodePerceptionPredicate,
+)
+from adam.semantics import (
+    SemanticNode,
+    Concept,
+    AffordanceConcept,
+    AffordanceSemanticNode,
+    ActionSemanticNode,
+)
+
+
+class AbstractAffordanceTemplateLearner(AbstractTemplateLearner, ABC):
+    def _can_learn_from(
+        self, language_perception_semantic_alignment: LanguagePerceptionSemanticAlignment
+    ) -> bool:
+        return (
+            len(
+                language_perception_semantic_alignment.perception_semantic_alignment.semantic_nodes
+            )
+            > 1
+        )
+
+    def _preprocess_scene(
+        self, perception_semantic_alignment: PerceptionSemanticAlignment
+    ) -> PerceptionSemanticAlignment:
+        return perception_semantic_alignment
+
+    def _candidate_templates(
+        self, language_perception_semantic_alignment: LanguagePerceptionSemanticAlignment
+    ) -> AbstractSet[SurfaceTemplateBoundToSemanticNodes]:
+        rtr = []
+        semantic_nodes = (
+            language_perception_semantic_alignment.perception_semantic_alignment.semantic_nodes
+        )
+        for action_node in semantic_nodes:
+            if isinstance(action_node, ActionSemanticNode):
+                for slot, object_node in action_node.slot_fillings.items():
+                    rtr.append(
+                        SurfaceTemplateBoundToSemanticNodes(
+                            surface_template=SurfaceTemplate(
+                                elements=(slot.name, action_node.concept.debug_string),
+                                language_mode=self._language_mode,
+                            ),
+                            slot_to_semantic_node={SLOT1: object_node},
+                        )
+                    )
+
+        return immutableset(rtr)
+
+    def _enrich_post_process(
+        self,
+        perception_graph_after_matching: PerceptionGraph,
+        immutable_new_nodes: AbstractSet[SemanticNode],
+    ) -> Tuple[PerceptionGraph, AbstractSet[SemanticNode]]:
+        digraph = perception_graph_after_matching.copy_as_digraph()
+        for node in immutable_new_nodes:
+            if (
+                isinstance(node, AffordanceSemanticNode)
+                and len(node.slot_fillings.values()) == 1
+            ):
+                # Ideally I'd like this weight to be the graph match percentage, but we don't have that information
+                # Available to us here and refactoring to have that information is out of scope
+                affordance_property = CategoricalNode(
+                    label="affordance", value=node.concept.debug_string, weight=1.0
+                )
+                digraph.add_node(affordance_property)
+                # This edge might not handle temporal encoding correctly?
+                digraph.add_edge(
+                    list(node.slot_fillings.values())[0],
+                    affordance_property,
+                    label=HAS_PROPERTY_LABEL,
+                )
```

Comment:
Right, for dynamic graphs we have to wrap this in a temporally scoped edge label, assuming we're continuing with the old representation. This code always uses an unwrapped label which breaks edge label consistency for that approach to dynamic scenes.

I don't see a good way of handling scope while keeping things in scope (ha), so maybe we want to use either (the equivalents of) `BEFORE` if we're being conservative or `ENTIRE_SCENE` otherwise. `BEFORE` seems like a safe bet -- the affordance we're learning has to hold `BEFORE` because otherwise the action couldn't have taken place in the scenes we learned it from. `ENTIRE_SCENE` is stronger but may not make sense for some affordances. Say we have an `edible`/`slot2_slot1_eat_slot2` affordance, in a scene `a person eats an apple` -- the affordance no longer logically applies after the eating. In the symbolic setting the object exists both before and after, if I recall correctly, which means we also can't couple it to the object's scope (though I don't think nodes even have temporal scope?).

I see approaches that might be better but I think they're out of scope. One better approach might be to learn the affordance patterns on the `BEFORE` scope, then match on each temporal scope separately. But that sounds complicated enough to be out of scope. An extension would be to learn patterns for each scope separately, but that seems even more complicated. I think we'd have to deal with chained actions (e.g. "I put the box on the table, then I put the box on the floor") to learn each scope separately, which is definitely out of scope.

On a related note, I think we should create an issue about how to handle temporal scopes for simulated action samples and link to that from here. We can plan to address this piece of code once we know how we will represent the samples as temporally scoped graphs. I think (correct me if I'm wrong) we haven't yet defined how we will translate ASU's output (which might? provide many frames of features, more than 2 or 3) into temporally scoped graphs, which complicates this. I think this will be clearer once we action samples to work with, so ideally we would wait on that.

Reply:
So I'd think in general the affordances should be able to be identified regardless of temporal scope, they are essentially 'not temporally determined' (or well shouldn't be) except for when used as a property. (e.g. an 'apple' is only edible before the 'eats' action because afterwards the 'apple' objects ceases to exist (if it still existed it would still be edible). The concern here with adding the node in should probably be safe to use `ENTIRE_SCENE` in our current configuration but I'm fine being conservative with `BEFORE`.  Ideally we'd just want to mirror the temporal scope used for 'object permanence' but I'm not quite sure we represent that.

I'll follow up with Temporal Scope in an issue.

Reply:
Right, I think I agree about not being temporally determined.

I weakly favor `BEFORE` because of `eats` but either scope works fine.